pallets/asset-index/src/lib.rs

```diff
@@ -215,19 +235,108 @@ pub mod pallet {
                 Error::<T>::MinimumRedemption
             );
 
-            let deposit = Self::index_token_balance(&caller);
-            ensure!(deposit >= amount, Error::<T>::InsufficientDeposit);
+            let free_balance = T::IndexToken::free_balance(&caller);
+            T::IndexToken::ensure_can_withdraw(
+                &caller,
+                amount,
+                WithdrawReasons::TRANSFER,
+                free_balance.saturating_sub(amount),
+            )?;
 
             let fee = T::WithdrawalFee::withdrawal_fee(amount);
-            let _redeem = amount
+            let redeem = amount
                 .checked_sub(&fee)
-                .ok_or(Error::<T>::InsufficientDeposit)?;
+                .ok_or(Error::<T>::InsufficientDeposit)?
+                .into();
+
+            // NOTE: the ratio of a liquid asset `a` is determined by `sum(nav_asset) / nav_a`
+            let mut liquid_assets_vol = T::Balance::zero();
+            let mut asset_prices = Vec::new();
+            for (asset, holding) in Holdings::<T>::iter().filter(|(_, holding)| holding.is_liquid())
+            {
+                let price = T::PriceFeed::get_price(asset)?;
+                let vol = Self::calculate_volume(holding.units, &price)?;
+                liquid_assets_vol = liquid_assets_vol
+                    .checked_add(&vol)
+                    .ok_or(Error::<T>::NAVOverflow)?;
+                asset_prices.push((price, vol));
+            }
 
-            // TODO calculate the distribution of assets
+            for (price, vol) in &mut asset_prices {
+                let ratio = Ratio::checked_from_rational(liquid_assets_vol.into(), (*vol).into())
+                    .ok_or(Error::<T>::NAVOverflow)?;
+                // overwrite the value with the units the user gets for that asset
+                *vol = ratio
+                    .checked_mul_int(redeem)
+                    .and_then(|units| price.reciprocal_volume(units))
+                    .ok_or(Error::<T>::AssetVolumeOverflow)
+                    .and_then(|units| {
+                        units.try_into().map_err(|_| Error::<T>::AssetUnitsOverflow)
+                    })?;
+            }
 
-            // TODO if total supply of any asset drops to 0 it gets removed from the index.
+            // update the index balance by burning all of the redeemed tokens and
+            // issuing new tokens for the tx fee into the treasury's account
+            let burned = T::IndexToken::burn(amount);
+            if let Err(burned) = T::IndexToken::settle(
+                &caller,
+                burned,
+                WithdrawReasons::TRANSFER,
+                ExistenceRequirement::KeepAlive,
+            ) {
+                // TODO can this even happen after ensure_can_withdraw?
```

Comment:
I think it would be too strict if we burn balances on the failing of `ensure_can_withdraw`, users just paid the tx fee.

To be confirmed, the design of the tx fee in this method exists along with the custom transaction fee calculated by the logic of substrate framework ( the fee `T::Weight`  determines )?

Reply:
that's a good point, I have to evaluate this.